catfs/db/database_badger.go

```diff
@@ -56,17 +51,60 @@ func NewBadgerDatabase(path string) (*BadgerDatabase, error) {
 				return
 			}
 
-			if err := db.RunValueLogGC(0.5); err != nil {
-				if err != badger.ErrNoRewrite {
-					log.WithError(err).Warnf("badger gc failed")
-				}
+			err := bdb.runGC()
+			if err != nil {
+				log.WithError(err).Error("badger GC failed")
 			}
 		}
 	}()
 
 	return bdb, nil
 }
 
+func (bdb *BadgerDatabase) runGC() error {
+	opts := bdb.db.Opts()
+	bdb.mu.Lock()
+	defer bdb.mu.Unlock()
+	// In large DB, GC will happen automatically, because compaction will find garbage
+	// but we are to small and compactors do not run (150 MB is small).
+	// So we need to run Flatten
+	bdb.db.Flatten(5)
+	// Very likely Flatten will not do much because the hard coded priority is too small.
+	// At this point, we hope that there is something for GC
+	var errGC error
+	var success = false
+	log.Debugf("Performing GC for badger DB in %s", opts.Dir)
+	for errGC == nil {
+		errGC = bdb.db.RunValueLogGC(0.5) // cleans DB online and it is safe to rerun on success
+		if errGC == nil && !success {
+			success = true
+		}
+	}
+	if success {
+		log.Debugf("Cleaned some garbage for DB in %s", opts.Dir)
+		return nil
+	}
+	// Now we have a dilemma: we could trust badger GC mechanism and stop here.
+	// But unfortunately for our typical size (even as high as 150 MB)
+	// compaction, even with Flatten(), does not kick in.
+	// The only way to truly force compaction (to update stats for GC) is to close DB
+	// see Note in https://github.com/dgraph-io/badger/issues/767#issuecomment-485713746
+	// After Close() the GC on a next run will have updated statistic
+	// Actually even Close() does not guaranteed success, it requires more than a minute
+	// to update stats after DB was modified. But eventually GC stats will be ready.
+	err := bdb.db.Close()
+	if err != nil {
+		log.Fatalf("Could not close DB in %s", opts.Dir)
```

Comment:
I generally see calling ``log.Fatalf`` as bad practice, especially if it's not in quick & dirty programs and especially in library code (I have a story to tell here). I'm not fully sure what the proper error handling here is, but we could enter a retry loop when fail to close or open the database.

Reply:
I tried to mitigate with 91d48265. Error in `Close` is not fatal, we can try later. If we cannot open, then we do ten attempts with timeouts and then Fail. We cannot run without DB. Also, fixed Close which did not check for an active transaction. GC could be done or running system, but Close definitely not.

I would love to hear this story :) I was always thinking that if something going wrong the program should faill with error message so user would send an angry bug report and we fix the problem. If we are running industrial process, then the story is different.

Reply:
> Error in Close is not fatal

Is it? What if it was partially closed and we're left with a zombie db conn?

> I would love to hear this story :)  If we are running industrial process, then the story is different.

Story is: We once used a [gpio library](https://github.com/brian-armstrong/gpio/pull/16/files) to check some pins on our device. Sadly I didn't look into the library implementation, so that fault is on me, but their error handling was "log & exit". Which is a bad strategy as a library. I noticed this first when the program didn't first on my work machine (obvious cause was that the gpio was not available there), but at some point we made the grave mistake to update the Go version to 1.14 and all hell broke loose. All services reading this pin suddenly crashed randomly after the update. And why was that? Because since 1.14 ``select()`` might return ``EINTR`` (rightly so) when it's execution was interrupted by the scheduler, which might happen at any point. The gpio library was not prepared for this and just jumped off the cliff by calling ``os.Exit(1)``. Fix was easy, but what we can learn from this:

* Never ever ``os.Exit()`` in library code.
* Check the quality of the library you're using.
* Error handling is hard.

> I was always thinking that if something going wrong the program should faill with error message so user would send an angry bug report and we fix the problem.

That's fine for interactive programs. Long running programs (and brig daemon counts as one) should do this only as last resort since the error gets buried in some logs (in the best case). But not being able to open the main database is one such case, so https://github.com/sahib/brig/commit/91d482653845748445bcb0e91335de793bc09acc is fine with me.

Reply:
I see now why it was a painful experience. Indeed `fail` is super special and should be used as a last line of defense.